### PR TITLE
Add `flake8-bugbear` as plugin to pre-commit

### LINKED ISSUE #6662: Add `flake8-bugbear` as additional `flake8` plugin
`flake8-bugbear` ([https://github.com/PyCQA/flake8-bugbear](https://github.com/PyCQA/flake8-bugbear)) is a plugin for `flake8` that contains additional rules aimed at finding likely bugs and design problems in the code. Notable rules include:
* B006: Do not use mutable data structures for argument defaults.
* B018: Found useless expression.

At the time of this writing, adding `flake8-bugbear` spits out 166 problems, most of which are in tests. If we decide to go with `flake8-bugbear`, we could ignore it for tests and clean those up over time before flipping the switch.

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #6662 

~Blocked by:~
* ~#6810~
* ~#6811~
* ~#6812~
* ~#6813~
* ~#6814~
* ~#6815~

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by hendrikmakait** [comment]
Should we also include **B904**?

> **B904:** Within an except clause, raise exceptions with raise ... from err or raise ... from None to distinguish them from errors in exception handling. See [the exception chaining tutorial](https://docs.python.org/3/tutorial/errors.html#exception-chaining) for details.

**Comment by fjetter** [comment]
> Should we also include B904?

My suggestion would be to enable whatever works right now, disable/ignore everything else, get the PR merged asap to avoid regressions. We can enable/disable more error codes in a follow up

File: distributed/tests/test_client.py [hunk 1]
```diff
@@ -3896,10 +3896,10 @@ def test_threaded_get_within_distributed(c):
 
     for get in [dask.local.get_sync, dask.multiprocessing.get, dask.threaded.get]:
 
-        def f():
+        def f(get):
```
Comment (hendrikmakait) [inline distributed/tests/test_client.py:3899]: B023

File: distributed/tests/test_client.py [hunk 2]
```diff
@@ -4017,8 +4017,8 @@ async def test_serialize_future(s, a, b):
     result = await future
 
     for ci in (c1, c2):
-        for ctxman in ci.as_current, lambda: temp_default_client(ci):
-            with ctxman():
+        for ctxman in ci.as_current, temp_default_client:
```
Comment (hendrikmakait) [inline distributed/tests/test_client.py:4020]: B023

File: distributed/worker.py [hunk 3]
```diff
@@ -1542,7 +1542,7 @@ async def close(  # type: ignore
             if executor is utils._offload_executor:
                 continue  # Never shutdown the offload executor
 
-            def _close(wait):
+            def _close(executor, wait):
```
Comment (hendrikmakait) [inline distributed/worker.py:1545]: B023

**Review by hendrikmakait — COMMENTED** [review]
Restructured some code that triggered 
> **B023:** Functions defined inside a loop must not use variables redefined in the loop, because [late-binding closures are a classic gotcha](https://docs.python-guide.org/writing/gotchas/#late-binding-closures).

File: distributed/tests/test_client.py [hunk 4]
```diff
@@ -4017,8 +4017,8 @@ async def test_serialize_future(s, a, b):
     result = await future
 
     for ci in (c1, c2):
-        for ctxman in ci.as_current, lambda: temp_default_client(ci):
-            with ctxman():
+        for ctxman in lambda ci: ci.as_current(), lambda ci: temp_default_client(ci):
```
Comment (hendrikmakait) [inline distributed/tests/test_client.py:4020]: B023

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -28,6 +30,10 @@ repos:
     hooks:
       - id: flake8
         language_version: python3
+        additional_dependencies:
+        # NOTE: autoupdate does not pick up flake8-bugbear since it is a transitive
+        #  dependency. Make sure to update flake8-bugbear manually on a regular basis.
+          - flake8-bugbear==22.7.1
```
Comment (hendrikmakait) [inline .pre-commit-config.yaml:36]: Pinned with added comments because `autoupdate` doesn't help us here.

**Comment by hendrikmakait** [comment]
CI flake: #6860

**Review by fjetter — APPROVED** [review]

**Pull request merged**